### PR TITLE
Cfy 3691 make all internal rest calls secure

### PR DESCRIPTION
**Pull request by noakup**

### CODE REVIEW

File: cloudify/constants.py [hunk 1]
```diff
@@ -41,3 +41,15 @@
 
 BROKER_PORT_NO_SSL = 5672
 BROKER_PORT_SSL = 5671
+CLOUDIFY_USERNAME_ENV = 'CLOUDIFY_USERNAME'
+CLOUDIFY_PASSWORD_ENV = 'CLOUDIFY_PASSWORD'
```
Comment (dankilman) [inline cloudify/constants.py]: 1) Need to decide if exposing the password as an environment variable to all operations is OK
2) Need to handle special characters in password in bash/batch scripts

File: cloudify/context.py [hunk 2]
```diff
@@ -97,6 +97,8 @@ def __init__(self, ctx=None):
 
         self.blueprint = BlueprintContext(self._context)
         self.deployment = DeploymentContext(self._context)
+        security_context_dict = self._context.get('security_context', {})
+        self.security_ctx = SecurityContext(**security_context_dict)
```
Comment (dankilman) [inline cloudify/context.py]: rename to `security_context`

File: cloudify/context.py [hunk 3]
```diff
@@ -244,6 +246,58 @@ def id(self):
         return self._context.get('deployment_id')
 
 
+class SecurityContext(object):
+
+    def __init__(self, security_enabled=None, ssl_enabled=None,
+                 verify_ssl_certificate=None, cloudify_username=None,
+                 cloudify_password=None):
+        self.is_security_enabled = security_enabled
```
Comment (dankilman) [inline cloudify/context.py]: save dict on self and implement `properties` with `dict.get...`

File: cloudify/context.py [hunk 4]
```diff
@@ -244,6 +246,58 @@ def id(self):
...
+        """The username of the currently active user"""
+        return self.cloudify_username
+
+    @property
+    def password(self):
+        """The password of the currently active user"""
+        return self.cloudify_password
+
+    @property
+    def properties(self):
```
Comment (dankilman) [inline cloudify/context.py]: consider removing this property and building this dict where it is used

File: cloudify/context.py [hunk 5]
```diff
@@ -658,7 +712,7 @@ def bootstrap_context(self):
         :rtype: BootstrapContext
         """
         if self._bootstrap_context is None:
-            context = self._endpoint.get_bootstrap_context()
+            context = self._endpoint.get_bootstrap_context(self.security_ctx)
```
Comment (dankilman) [inline cloudify/context.py]: use ctx from self

File: cloudify/manager.py [hunk 6]
```diff
@@ -107,13 +108,36 @@ def relationships(self):
         return self._relationships
 
 
-def get_rest_client():
+# This method is wrong in so many ways... should the credentials be passed to
+# it or are they expected to be env vars, or both?
+# what about the rest of the settings?
+# since it's using code a lot of code from utils, shouldn't it be in utils
+# actually?
+def get_rest_client(security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 7]
```diff
@@ -107,13 +108,36 @@ def relationships(self):
...
+    if security_ctx.security_enabled:
+        if security_ctx.ssl_enabled:
+            rest_port = constants.SECURED_REST_PORT
+            protocol = constants.SECURED_PROTOCOL
+
+    headers = utils.get_auth_header(security_ctx.username,
+                                    security_ctx.password)
+    if security_ctx.verify_ssl_certificate:
+        trust_all = False
+        cert_path = constants.SSL_CERT_PATH
```
Comment (dankilman) [inline cloudify/manager.py]: 1) Only use constant if not explicit value
2) Need to copy cert to agents (not so clear how to do so for init_script/provided installation methods)

File: cloudify/manager.py [hunk 8]
```diff
@@ -191,14 +215,14 @@ def get_blueprint_resource(blueprint_id, resource_path):
     return get_resource(resource_path, base_url=base_url)
 
 
-def get_node_instance(node_instance_id):
+def get_node_instance(node_instance_id, security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 9]
```diff
@@ -209,13 +233,13 @@ def get_node_instance(node_instance_id):
                         relationships=instance.relationships)
 
 
-def update_node_instance(node_instance):
+def update_node_instance(node_instance, security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 10]
```diff
@@ -223,12 +247,12 @@ def update_node_instance(node_instance):
         version=node_instance.version)
 
 
-def get_node_instance_ip(node_instance_id):
+def get_node_instance_ip(node_instance_id, security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 11]
```diff
@@ -248,26 +272,26 @@ def get_node_instance_ip(node_instance_id):
 # TODO: some nasty code duplication between these two methods
 
 
-def update_execution_status(execution_id, status, error=None):
+def update_execution_status(execution_id, status, security_ctx, error=None):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 12]
```diff
@@ -248,26 +272,26 @@ def get_node_instance_ip(node_instance_id):
...
 
     :returns: The updated status
     """
-    client = get_rest_client()
+    client = get_rest_client(security_ctx)
     return client.executions.update(execution_id, status, error)
 
 
-def get_bootstrap_context():
+def get_bootstrap_context(security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/manager.py [hunk 13]
```diff
@@ -248,26 +272,26 @@ def get_node_instance_ip(node_instance_id):
...
     """Read the manager bootstrap context."""
-    client = get_rest_client()
+    client = get_rest_client(security_ctx)
     context = client.manager.get_context()['context']
     return context.get('cloudify', {})
 
 
-def get_provider_context():
+def get_provider_context(security_ctx):
```
Comment (dankilman) [inline cloudify/manager.py]: make optional

File: cloudify/workflows/local.py [hunk 14]
```diff
@@ -127,7 +128,8 @@ def execute(self,
             'task_retries': task_retries,
             'task_retry_interval': task_retry_interval,
             'subgraph_retries': subgraph_retries,
-            'local_task_thread_pool_size': task_thread_pool_size
+            'local_task_thread_pool_size': task_thread_pool_size,
+            'security_ctx': security_ctx
```
Comment (dankilman) [inline cloudify/workflows/local.py]: no need to pass in local workflow context

File: cloudify/workflows/local.py [hunk 15]
```diff
@@ -98,6 +98,7 @@ def evaluate_functions(self, payload, context):
 
     def execute(self,
                 workflow,
+                security_ctx,
```
Comment (dankilman) [inline cloudify/workflows/local.py]: remove from signature

File: cloudify/workflows/workflow_context.py [hunk 16]
```diff
@@ -621,7 +625,8 @@ def execute_task(self,
         cloudify_context = self._build_cloudify_context(
             task_id,
             task_name,
-            node_context)
+            node_context,
+            self.security_ctx.properties)
```
Comment (dankilman) [inline cloudify/workflows/workflow_context.py]: remove `properties` and pass in `_sec_context`

File: cloudify/workflows/workflow_context.py [hunk 17]
```diff
@@ -1100,14 +1111,26 @@ def _derive(property_name):
...
                 if len(runtime_props) == 0:
-                    host_node_instance = get_node_instance(host_id)
-                    cloudify_agent = host_node_instance.runtime_properties.get(
-                        'cloudify_agent')
-                    if not cloudify_agent:
-                        raise exceptions.NonRecoverableError(
-                            'Missing cloudify_agent runtime information. '
-                            'This most likely means that the Compute node '
-                            'never started successfully')
+                    with open('/tmp/workflow_context.log', 'a') as log_file:
```
Comment (dankilman) [inline cloudify/workflows/workflow_context.py]: cleanup

File: cloudify/workflows/workflow_context.py [hunk 18]
```diff
@@ -807,19 +812,25 @@ def __init__(self, ctx):
...
-            rest = get_rest_client()
+            rest = get_rest_client(self._security_ctx)
             raw_nodes = rest.nodes.list(self.deployment.id)
             raw_node_instances = rest.node_instances.list(self.deployment.id)
 
         WorkflowNodesAndInstancesContainer.__init__(self, self, raw_nodes,
                                                     raw_node_instances)
 
+    @property
+    def security_ctx(self):
```
Comment (dankilman) [inline cloudify/workflows/workflow_context.py]: remove public property and make use of private in all places

File: cloudify/workflows/workflow_context.py [hunk 19]
```diff
@@ -404,6 +405,8 @@ class _WorkflowContextBase(object):
 
     def __init__(self, ctx, remote_ctx_handler_cls):
         self._context = ctx = ctx or {}
+        security_ctx_dict = self._context.get('security_ctx', {})
+        self._security_ctx = SecurityContext(**security_ctx_dict)
```
Comment (dankilman) [inline cloudify/workflows/workflow_context.py]: need to handle system wide workflows use case (get_rest_client)

**Comment by noakup** [comment]
very similar to #158, closing but not deleting branch until #158 is merged.

**Pull request closed without merging**